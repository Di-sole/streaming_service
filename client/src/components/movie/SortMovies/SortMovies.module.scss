.container {
    // width: 1216px;
    // margin-left: auto;
    // margin-right: auto;

    &__sortingControl {
        display: flex;
        align-items: center;
        width: 100%;
    }

    &__sorting {
        position: relative;
        overflow: hidden;

        &_active {
            overflow: visible;

            .container__sortArrow {
                &::before {
                    transform: rotate(180deg);
                }
            }

            .container__dropdown {
                pointer-events: auto;
            }
        }
    }

    &__sortingPanel {
        display: flex;
        align-items: center;
        cursor: pointer;
    }

    &__sortIcon {
        display: flex;
        width: 16px;
        height: 16px;
        margin-right: 8px;

        font-size: 16px;
        line-height: 16px;

        &::before {
            color: #fff;

            content: url('../../../../public/icons/sortLines.svg');
        }
    }

    &__sortParameter {
        height: 20px;

        font-size: 15px;
        line-height: 20px;
        font-weight: 500;
        overflow: hidden;
        white-space: nowrap;
        text-overflow: ellipsis;

        color: #fff;
    }

    &__sortArrow {
        margin-left: 6px;

        &::before {
            display: inline-block;
            content: url('../../../../public/icons/arrows/arrow_down.svg');
        }
    }

    &__dropdown {
        min-width: 250px;

        position: absolute;
        top: 32px;
        z-index: 2;

        display: inline-block;
        padding-top: 8px;
        padding-bottom: 12px;
        background-color: #1f1b2e;
        box-shadow: 0 20px 40px rgba(7, 5, 14, 0.4);
        pointer-events: auto;
    }

    &__itemTitle {
        margin-left: 4px;
        margin-top: 4px;
        margin-bottom: 12px;
        padding: 0 16px;
        font-size: 15px;
        line-height: 20px;
        font-weight: 400;
        color: #a5a1b2;
    }

    &__dropdownItem {
        display: block;
        position: relative;
        padding-left: 20px;
        padding-right: 16px;
        padding-top: 12px;
        padding-bottom: 12px;

        background-color: transparent;

        white-space: nowrap;
        cursor: pointer;

        &_selected {
            .container__itemText {
                color: #fff;
            }

            .container__itemStripe {
                display: flex;
            }
        }

        &:hover {
            background-color: #312b45;
        }
    }

    &__itemStripe {
        display: none;
        position: absolute;
        left: 0;
        top: 0;
        bottom: 0;
        width: 4px;

        background-color: #ea003d;
    }

    &__itemText {
        display: inline-block;
        vertical-align: top;
        max-height: 20px;

        font-size: 15px;
        line-height: 20px;
        font-weight: 500;
        color: rgba(255, 255, 255, 0.72);

        overflow: hidden;
        white-space: nowrap;
        text-overflow: ellipsis;
    }
}

@media screen and (min-width: 600px) {
    .container {
        &__sortingControl {
            margin-top: 24px;
            // padding-left: 12px;
            padding-right: 12px;
        }
    }
}
