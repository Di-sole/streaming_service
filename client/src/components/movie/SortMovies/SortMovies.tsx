import { useEffect, useState } from 'react';
import styles from './SortMovies.module.scss';
import { useAppDispatch, useAppSelector } from '@/store/hooks/redux';
import { getSortedMovies } from '@/utils/moviesHelpers';
import IMovies from '@/models/IMovies';
import { useTranslation } from 'react-i18next';
import { useRouter } from 'next/router';

export interface ISortParams {
    userRates: string;
    rating: string;
    releaseDate: string;
    movieName: string;
}

const SortMovies = ({ filteredMovies }: { filteredMovies: IMovies[] }) => {
    const { t } = useTranslation('moviesPage');
    const { locale } = useRouter();
    const dispatch = useAppDispatch();
    const [isDropdownShown, setIsDropdownShown] = useState(false);

    const sortParams: ISortParams = {
        userRates: t('sorting.userRates'),
        rating: t('sorting.rating'),
        releaseDate: t('sorting.releaseDate'),
        movieName: t('sorting.name'),
    };
    const [sortingParameter, setSortingParameter] = useState(sortParams.userRates);

    //const sortingParamNames = ['По количеству оценок', 'Рейтингу', 'Дате выхода', 'Алфавиту'];
    const sortingParamNames = [
        t('sorting.userRates'),
        t('sorting.rating'),
        t('sorting.releaseDate'),
        t('sorting.name'),
    ];

    useEffect(() => {
        getSortedMovies(sortParams, sortingParameter, [...filteredMovies], locale!, dispatch);
    }, [sortingParameter]);

    const handleSortTypeSelection = ({ currentTarget }: React.MouseEvent<HTMLDivElement>) => {
        setSortingParameter(currentTarget.textContent!);
        currentTarget
            .parentElement!.querySelector(`.${styles.container__dropdownItem_selected}`)
            ?.classList.remove(styles.container__dropdownItem_selected);
        currentTarget.classList.add(styles.container__dropdownItem_selected);
    };

    const renderMenuItem = (paramName: string) => {
        return (
            <div className={styles.container__dropdownItem} key={paramName} onClick={handleSortTypeSelection}>
                <div className={styles.container__itemStripe}></div>
                <div className={styles.container__itemText}>{paramName}</div>
            </div>
        );
    };

    const selectedParam =
        sortingParameter === sortParams.userRates
            ? sortingParameter
            : `${t('sorting.by')} ${sortingParameter.slice(0, 1).toLowerCase()}${sortingParameter.slice(1)}`;

    return (
        <section>
            <div className={styles.container}>
                <div className={styles.container__sortingControl}>
                    <div
                        className={[
                            styles.container__sorting,
                            isDropdownShown ? styles.container__sorting_active : null,
                        ].join(' ')}
                        onClick={() => setIsDropdownShown(!isDropdownShown)}
                    >
                        <div className={styles.container__sortingPanel}>
                            <div className={styles.container__sortIcon}></div>
                            <div className={styles.container__sortParameter}>{selectedParam}</div>
                            <div className={styles.container__sortArrow}></div>
                        </div>
                        <div className={styles.container__dropdown}>
                            <div className={styles.container__itemTitle}>Сортировать</div>
                            {Object.values(sortParams).map((param) => renderMenuItem(param))}
                        </div>
                    </div>
                </div>
            </div>
        </section>
    );
};
export default SortMovies;
