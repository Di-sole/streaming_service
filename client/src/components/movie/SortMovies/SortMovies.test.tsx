import { render, screen } from '@testing-library/react';
import SortMovies from './SortMovies';
import { moviesTest } from '../../../testAsserts/testItems';
import * as reduxHooks from '../../../store/hooks/redux';

jest.mock('../../../store/hooks/redux');
const useDispatch = jest.spyOn(reduxHooks, 'useAppDispatch');
const mockDispatch = jest.fn();

jest.mock('next-i18next', () => ({
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => {}),
            },
        };
    },
}));

jest.mock('next/router', () => ({
    useRouter() {
        return {
            pathname: '',
        };
    },
}));

describe('SORTMOVIES SNAPSHOTS TESTS', () => {
    test('SortMovies should not be changed', () => {
        useDispatch.mockReturnValue(mockDispatch);
        render(<SortMovies filteredMovies={moviesTest} />);

        const sort = screen.getByTestId('sortMovies');
        expect(sort).toMatchSnapshot();
    });
});
