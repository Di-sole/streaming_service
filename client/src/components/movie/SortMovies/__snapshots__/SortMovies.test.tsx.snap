// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SORTMOVIES SNAPSHOTS TESTS SortMovies should not be changed 1`] = `
<section
  data-testid="sortMovies"
>
  <div
    class="container"
  >
    <div
      class="container__sortingControl"
    >
      <div
        class="container__sorting "
      >
        <div
          class="container__sortingPanel"
        >
          <div
            class="container__sortIcon"
          />
          <div
            class="container__sortParameter"
          >
            sorting.by sorting.userRates
          </div>
          <div
            class="container__sortArrow"
          />
        </div>
        <div
          class="container__dropdown"
        >
          <div
            class="container__itemTitle"
          >
            sorting.sort
          </div>
          <div
            class="container__dropdownItem"
          >
            <div
              class="container__itemStripe"
            />
            <div
              class="container__itemText"
            >
              sorting.by sorting.userRates
            </div>
          </div>
          <div
            class="container__dropdownItem"
          >
            <div
              class="container__itemStripe"
            />
            <div
              class="container__itemText"
            >
              sorting.rating
            </div>
          </div>
          <div
            class="container__dropdownItem"
          >
            <div
              class="container__itemStripe"
            />
            <div
              class="container__itemText"
            >
              sorting.releaseDate
            </div>
          </div>
          <div
            class="container__dropdownItem"
          >
            <div
              class="container__itemStripe"
            />
            <div
              class="container__itemText"
            >
              sorting.name
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
